### PR TITLE
internal/envoy: Make merge_slashes option configurable

### LINKED ISSUE #4339: support disabling Envoy's merge_slashes option
**What steps did you take and what happened:**
An upstream resource with URL path `/wiki/Mitou_Shoukan://Blood-Sign_-_Tập_1_Ending_1` becomes inaccessible when exposed via Contour - the path is normalized to `/wiki/Mitou_Shoukan:/Blood-Sign_-_Tập_1_Ending_1` in the upstream request.

**What did you expect to happen:**
The URL should be accessible.

**Anything else you would like to add:**
~~It seems this is due to https://github.com/projectcontour/contour/pull/988 enabling `normalize_path` for connection managers as a mitigation for https://github.com/envoyproxy/envoy/security/advisories/GHSA-xcx5-93pw-jw2w. Currently, this value is not configurable.~~

Edit: The cause is [`merge_slashes`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-merge-slashes) rather than `normalize_path`.

**Environment:**

- Contour version: 1.20.0
- Kubernetes version: (use `kubectl version`): N/A
- Kubernetes installer & version: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from `/etc/os-release`): N/A

### PR DESCRIPTION
**Pull request by mszabo-wikia**
#2406 enabled Envoy's non-standard `merge_slashes` path transformation
option[1] by default, which strips duplicate slashes from request URLs.
This can cause issues with upstream applications that do not use them
as an RFC 3986 delimiter,[2] so expose a configuration setting to toggle
this behavior. This toggle defaults to true to maintain compatibility
with existing installs.

Fixes #4339

---
[1] https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-merge-slashes
[2] https://datatracker.ietf.org/doc/html/rfc3986#section-2.2

### CODE REVIEW

**Comment by mszabo-wikia** [comment]
Looks like I missed some signature changes during rebase, looking at that now.

**Comment by mszabo-wikia** [comment]
With the changelog file name corrected, this should hopefully be ready now :)

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (skriss) [inline apis/projectcontour/v1alpha1/contourconfig.go:294]: I think this should be switched to `DisableMergeSlashes`, following a similar pattern as for the above field `DisableAllowChunkedLength`, with a default value of `false`. This way, Contour's default behavior doesn't change, and the user only needs to specify something if they want to change the default behavior. Otherwise, we run into problems where we can't distinguish between this field not being specified, and being explicitly specified as `false`.

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:338]: Assuming we switch to `DisableMergeSlashes`, this is where I'd apply the negation (just like the line just above for `DisableAllowChunkedLength`), so that the `xdscache_v3.ListenerConfig` still just has a `MergeSlashes` field.

File: pkg/config/parameters.go
Comment (skriss) [inline pkg/config/parameters.go:618]: Would also switch this one to `DisableMergeSlashes` with a default value of `false`. Otherwise, if this field is not specified in the config file, it's indistinguishable from it being set to `false` which would change the default behavior for users.

File: changelogs/unreleased/4363-mszabo-wikia-small.md
Comment (skriss) [inline changelogs/unreleased/4363-mszabo-wikia-small.md:2]:
```suggestion
a non-standard path transformation option to replace multiple consecutive slashes in an URL path with a single slash can now be disabled by setting the `DisableMergeSlashes` option in the Contour config file or ContourConfiguration custom resource.
```

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks for the PR @mszabo-wikia! I have one piece of feedback on the polarity of the config field, let's wait for @youngnick's input as well before switching everything just to make sure he agrees with me.

File: apis/projectcontour/v1alpha1/contourconfig.go
Comment (mszabo-wikia) [reply apis/projectcontour/v1alpha1/contourconfig.go:294]: Thanks, this sounds good. I thought applying the `kubebuilder` default would have sufficed for that but I see now that it only controls the generated example resources.

**Comment by mszabo-wikia** [comment]
Thanks for the feedback! I've amended the PR accordingly

**Review by skriss — APPROVED** [review]
Thanks for the updates @mszabo-wikia, LGTM! Let's wait for one more maintainer to review as well before merging.

**Review by sunjayBhatia — APPROVED** [review]
LGTM, fan of the field naming/polarity for the config for users 👍🏽

**Review by youngnick — APPROVED** [review]
This is great, thanks @mszabo-wikia !

**Comment by mszabo-wikia** [comment]
Thanks for the review and guidance! :)

**Pull request merged**
